<a href="https://colab.research.google.com/github/rroszczyk/Python/blob/master/okno/OKNO_1A.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [44]:
# przygotowanie katalogu roboczego do pracy :)
import requests

def pobierzPlik(url, nazwaPliku):
    req = requests.get(url, stream=True)
    with open(nazwaPliku, 'wb') as uchwytPliku:
        for chunk in req.iter_content():
            uchwytPliku.write(chunk)

nazwyPlikow = ['Sample_Weather_Data.csv']
for nazwa in nazwyPlikow:
  pobierzPlik("https://raw.githubusercontent.com/rroszczyk/Python/master/Pliki/"+nazwa, nazwa)

# Import bibliotek wymaganych do pracy z Pandas

In [2]:
import pandas as pd

In [3]:
pd.__version__

'2.2.2'

# Wczytanie danych

Aby załadować zbiór danych, mamy różne opcje w zależności od formatu źródłowego pliku. Tutaj mamy format tekstowy jako csv, więc użyjemy metody Pandas read_csv(), która odczyta csv i załaduje cały zbiór danych do DataFrame.


In [4]:
df = pd.read_csv(r'Sample_Weather_Data.csv')

Ogólną zasadą w każdym problemie uczenia maszynowego jest poznanie swoich danych, zrozumienie swoich danych. Wyciągnij z nich podstawowe informacje, takie jak:

- ile punktów danych lub obserwacji ma nasz zestaw danych?
- czy ma jakieś brakujące wartości?
- ile kolumn/cech ma nasz zestaw danych?
- jaki jest typ danych każdej kolumny?
i tak dalej.

Na podstawie tego podejmuje się decyzję o kolejnym kroku.

Możesz nazwać to dowolną nazwą, taką jak EDA (Exploratory Data Analysis); lub Understanding Data; lub Exploring Data; Performing Data Analysis itd.

Celem jest tutaj uzyskanie podstawowego zrozumienia zestawu danych i problemu, który próbujemy rozwiązać.

Nie tracąc czasu, zobaczmy, jakie metody udostępnił nam Python i Pandas.

# head()

Pierwszą rzeczą po załadowaniu danych jest sprawdzenie, jak wyglądają nasze dane. W tym celu musimy rzucić okiem na pierwsze kilka punktów danych. W tym celu mamy metodę zwaną head().

Metoda head() Pandasa jest używana do zwracania n górnych (domyślnie 5) wierszy ramki danych lub serii.

In [5]:
df.head()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog

Widzimy pierwszych 5 rekordów. Ponieważ nie podaliśmy żadnej wartości, domyślnie jest to 5, co zwraca pierwszych 5 rekordów. W ramkach danych Pandas indeksowanie zaczyna się od 0, więc mamy od 0 do 4, co daje łącznie 5.

In [6]:
df.head(8)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   
5  1/1/2012 5:00    -1.4              -3.3         87                9   
6  1/1/2012 6:00    -1.5              -3.1         89                7   
7  1/1/2012 7:00    -1.4              -3.6         85                7   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog  
5            6.4     101.27                   Fog  
6            6.4     101.29                   Fog  
7            8.0     101.26                   Fog

Otrzymaliśmy pierwszych 8 rekordów.

# tail()

Podobnie jak metoda `head()`, Pandas posiada metodę `tail()` do wyświetlania ostatnich n wierszy.

Metoda `tail()` w Pandas służy do zwracania ostatnich n (domyślnie 5) wierszy ramki danych lub szeregu.

In [7]:
df.tail()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather  
8779            9.7     100.13    Snow  
8780            9.7     100.03    Snow  
8781            4.8      99.95    Snow  
8782            9.7      99.91    Snow  
8783           11.3      99.89    Snow

Widzimy ostatnich 5 rekordów. Ponieważ tak jak poprzednio nie podaliśmy żadnej wartości, domyślnie jest to ponownie 5.

In [8]:
df.tail(2)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather  
8782            9.7      99.91    Snow  
8783           11.3      99.89    Snow

Tutaj widzimy ostatnie 2 rekordy.

Ostatni rekord ma indeks 8783.

# shape
Pokazuje całkowitą liczbę wierszy (punktów danych lub obserwacji) i całkowitą liczbę kolumn ramki danych.

Pandas udostępnia atrybuty `.size` i `.ndim` wraz z `.shape`, które służą do zwracania odpowiednio rozmiaru, wymiarów i kształtu ramek danych oraz szeregów.

In [9]:
df.shape

(8784, 8)

Zatem jest łącznie 8784 rekordów lub punktów danych z 8 kolumnami.

Stąd możemy od razu powiedzieć, że ramka danych jest dwuwymiarowa, a jej rozmiar to 8784 * 8 = 70272.

Te same informacje można sprawdzić za pomocą metod `.size` i `.ndim` w Pandas.

In [10]:
df.size

70272

In [11]:
df.shape[0] * df.shape[1]

70272

Jak widać, rozmiar to po prostu iloczyn liczby wierszy i kolumn, co można alternatywnie zapisać jako `df.shape[0] * df.shape[1]`.

In [12]:
df.ndim

2

Jak już wiedzieliśmy, nasza ramka danych jest dwuwymiarowa. Zwykle problemy uczenia maszynowego operują na dwuwymiarowych ramkach danych, a gdy pracujemy z obrazami (szczególnie w odcieniach szarości lub kolorowymi), mamy do czynienia z tablicami trójwymiarowymi. Zachęcam do zgłębienia tego tematu, jeśli jesteś zainteresowany.

A co jeśli musimy zweryfikować wymiar pojedynczej kolumny lub szeregu? Nie ma problemu, zróbmy to.

In [13]:
df['Wind Speed_km/h'].ndim

1

Gdy pracujemy z pojedynczymi kolumnami z ramki danych, są one konwertowane na szeregi (Series), a szeregi są jednowymiarowe, jak widzieliśmy powyżej.

Szereg (Series) w Pandas to jednowymiarowa, etykietowana tablica zdolna do przechowywania danych dowolnego typu (liczby całkowite, ciągi znaków, liczby zmiennoprzecinkowe, obiekty Pythona itp.). Należy pamiętać, że w przeciwieństwie do list w Pythonie, szereg zawsze będzie zawierał dane tego samego typu.

# index
Atrybut `index` dostarcza indeks lub zakres indeksu ramki danych.

W naszym przykładowym zbiorze danych wiemy, że indeks zaczął się od zera (0), co sprawdziliśmy za pomocą `.head()`, a ostatni wiersz ma indeks 8783, co sprawdziliśmy za pomocą metody `.tail()`.

To samo weryfikujemy za pomocą `.index`.

In [14]:
df.index

RangeIndex(start=0, stop=8784, step=1)

Ups, otrzymaliśmy RangeIndex, który mówi, że zaczął się od 0, zgadza się, ale jak to możliwe, że zatrzymuje się na 8784, podczas gdy wiemy, że ostatni indeksowany wiersz to 8783.

Tutaj jest haczyk: w zakresie indeksu ostatnia wartość jest wykluczona. Zatem, gdy mówi, że zatrzymuje się na 8784, rzeczywisty indeks kończy się na 8784 - 1 = 8783.

# columns
Następny jest atrybut `.columns`, który służy do pobrania wszystkich nazw kolumn w ramce danych.

In [15]:
df.columns

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather'],
      dtype='object')

Tutaj otrzymamy listę kolumn z ramki danych.

# dtypes

Atrybut `.dtypes` w Pandas służy do pobierania typu danych każdej kolumny w ramce danych (DataFrame).



In [16]:
df.dtypes

Date/Time            object
Temp_C              float64
Dew Point Temp_C    float64
Rel Hum_%             int64
Wind Speed_km/h       int64
Visibility_km       float64
Press_kPa           float64
Weather              object
dtype: object

Powyższy wynik pokazuje typ danych każdej kolumny w ramce danych.

A co, jeśli interesuje nas tylko jedna lub określony zestaw kolumn? Zobaczmy to również.

In [17]:
df['Wind Speed_km/h'].dtypes

dtype('int64')

Oto typ danych pojedynczej kolumny. Jako przykładu użyłem `Wind Speed_km/h`, ale można użyć dowolnej kolumny.

In [18]:
df[['Wind Speed_km/h', 'Weather']].dtypes

Wind Speed_km/h     int64
Weather            object
dtype: object

Sprawdzając typ danych wielu kolumn, upewnij się, że podajesz je w podwójnych nawiasach kwadratowych [[ ]], jak pokazano powyżej.

# unique()
Metoda `unique()` w Pandas służy do wyświetlania wszystkich unikalnych wartości dla danej kolumny. Pamiętaj, że metoda `unique()` może być zastosowana tylko do pojedynczej kolumny, a nie do całej ramki danych.

In [19]:
df['Wind Speed_km/h'].unique()

array([ 4,  7,  6,  9, 15, 13, 20, 22, 19, 24, 30, 35, 39, 32, 33, 26, 44,
       43, 48, 37, 28, 17, 11,  0, 83, 70, 57, 46, 41, 52, 50, 63, 54,  2])

In [20]:
df.Weather.unique()

array(['Fog', 'Freezing Drizzle,Fog', 'Mostly Cloudy', 'Cloudy', 'Rain',
       'Rain Showers', 'Mainly Clear', 'Snow Showers', 'Snow', 'Clear',
       'Freezing Rain,Fog', 'Freezing Rain', 'Freezing Drizzle',
       'Rain,Snow', 'Moderate Snow', 'Freezing Drizzle,Snow',
       'Freezing Rain,Snow Grains', 'Snow,Blowing Snow', 'Freezing Fog',
       'Haze', 'Rain,Fog', 'Drizzle,Fog', 'Drizzle',
       'Freezing Drizzle,Haze', 'Freezing Rain,Haze', 'Snow,Haze',
       'Snow,Fog', 'Snow,Ice Pellets', 'Rain,Haze', 'Thunderstorms,Rain',
       'Thunderstorms,Rain Showers', 'Thunderstorms,Heavy Rain Showers',
       'Thunderstorms,Rain Showers,Fog', 'Thunderstorms',
       'Thunderstorms,Rain,Fog',
       'Thunderstorms,Moderate Rain Showers,Fog', 'Rain Showers,Fog',
       'Rain Showers,Snow Showers', 'Snow Pellets', 'Rain,Snow,Fog',
       'Moderate Rain,Fog', 'Freezing Rain,Ice Pellets,Fog',
       'Drizzle,Ice Pellets,Fog', 'Drizzle,Snow', 'Rain,Ice Pellets',
       'Drizzle,Snow,Fog', 

Pracując z kolumnami, możemy odwoływać się do nich na dwa sposoby:
* df.Weather --> Gdy nazwa kolumny nie zawiera spacji.
* df['Weather'] --> zawsze zalecane. Zwiększa czytelność.

Powyższa metoda `unique()` zwróciła wszystkie unikalne wartości dla danej kolumny.

# nunique()
Metoda `.nunique()` w Pandas służy do pokazania całkowitej liczby unikalnych wartości w każdej kolumnie. Jak wspomniano, można ją zastosować do całej ramki danych, jak również do pojedynczej kolumny lub podzbioru kolumn.

In [21]:
df.nunique()

Date/Time           8784
Temp_C               533
Dew Point Temp_C     489
Rel Hum_%             83
Wind Speed_km/h       34
Visibility_km         24
Press_kPa            518
Weather               50
dtype: int64

Użyłem metody `nunique` na całej ramce danych. Teraz zobaczmy, jak użyć jej na pojedynczej kolumnie i na podzbiorze kolumn.

In [22]:
df.Weather.nunique()

50

In [23]:
df[['Wind Speed_km/h', 'Weather']].nunique()

Wind Speed_km/h    34
Weather            50
dtype: int64

Powyższe dwa fragmenty kodu pokazują, że całkowita liczba unikalnych wartości w kolumnie 'Weather' wynosi 50, a przy sprawdzaniu jako podzbiór kolumn, wynik jest taki sam, razem z wynikami dla wszystkich podanych kolumn.

# count()
Metoda `count()` w Pandas służy do pokazania całkowitej liczby nie-nullowych wartości w każdej kolumnie.

Metoda `count()` może być zastosowana zarówno do pojedynczej kolumny, jak i do całej ramki danych.

In [24]:
df.count()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

Z tego możemy łatwo wywnioskować, że w naszej ramce danych nie ma żadnych wartości null ani NaN. Ramka zawiera 8784 wierszy, a każda kolumna również ma 8784 wartości (niezależnie od tego, czy są zduplikowane, czy unikalne).

In [25]:
df.Weather.count()

np.int64(8784)

In [26]:
df[['Wind Speed_km/h', 'Weather']].count()

Wind Speed_km/h    8784
Weather            8784
dtype: int64

Jak wspomniano, metoda `count()` może być również używana na pojedynczej kolumnie lub na podzbiorze kolumn, co zostało pokazane powyżej.

# isnull() i notnull()
Podczas tworzenia ramki danych z pliku CSV, wiele pustych kolumn jest importowanych jako wartości null do ramki danych, co później stwarza problemy podczas operacji na tej ramce danych.

Czasami sam nasz zbiór danych zawiera brakujące punkty danych, które po zaimportowaniu do ramki danych zostaną przekonwertowane na wartości null lub NaN.

Metody `isnull()` i `notnull()` w Pandas służą do sprawdzania i zarządzania wartościami NULL w ramce danych.

In [27]:
df.isnull()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0         False   False             False      False            False   
1         False   False             False      False            False   
2         False   False             False      False            False   
3         False   False             False      False            False   
4         False   False             False      False            False   
...         ...     ...               ...        ...              ...   
8779      False   False             False      False            False   
8780      False   False             False      False            False   
8781      False   False             False      False            False   
8782      False   False             False      False            False   
8783      False   False             False      False            False   

      Visibility_km  Press_kPa  Weather  
0             False      False    False  
1             False      False    False  
2             False      False    False  
3             False      False    False  
4             False      False    False  
...             ...        ...      ...  
8779          False      False    False  
8780          False      False    False  
8781          False      False    False  
8782          False      False    False  
8783          False      False    False  

[8784 rows x 8 columns]

`isnull()` zwróci True, jeśli w wartości znajduje się jakakolwiek wartość NaN, w przeciwnym razie zwróci False.

Jednak przeglądanie tak dużego zbioru danych z wartościami True lub False byłoby bardzo żmudne. Aby to uprościć, możemy zagregować te dane i sprawdzić, czy istnieją jakiekolwiek wartości null, używając `sum()` na wyniku `isnull()`, jak poniżej.

In [28]:
df.isnull().sum()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

Z metody `counts()` już wiedzieliśmy, że nasz zbiór danych nie ma wartości null, co potwierdziliśmy również za pomocą metody `isnull().sum()`.

Przeciwieństwem `isnull()` jest metoda `notnull()`, która zwróci True dla każdej wartości NIE-NULL i False dla wartości null.

Tutaj również możemy użyć funkcji agregującej `.sum()`, aby uzyskać liczbę.

In [29]:
df.notnull()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0          True    True              True       True             True   
1          True    True              True       True             True   
2          True    True              True       True             True   
3          True    True              True       True             True   
4          True    True              True       True             True   
...         ...     ...               ...        ...              ...   
8779       True    True              True       True             True   
8780       True    True              True       True             True   
8781       True    True              True       True             True   
8782       True    True              True       True             True   
8783       True    True              True       True             True   

      Visibility_km  Press_kPa  Weather  
0              True       True     True  
1              True       True     True  
2              True       True     True  
3              True       True     True  
4              True       True     True  
...             ...        ...      ...  
8779           True       True     True  
8780           True       True     True  
8781           True       True     True  
8782           True       True     True  
8783           True       True     True  

[8784 rows x 8 columns]

In [30]:
df.notnull().sum()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

# value_counts()

Pandas `value_count()` to kolejna metoda służąca do pokazywania wszystkich unikalnych wartości wraz z ich liczbą dla danej kolumny. Ta metoda jest stosowana tylko do pojedynczych kolumn.

Wynikowy obiekt będzie posortowany w porządku malejącym, tak aby pierwszy element był najczęściej występującym elementem. Domyślnie wyklucza wartości NA.

In [31]:
df['Weather'].value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                              7
Freezing Drizzle,Fog                          6
Snow,Ice Pellets                              6
Snow,Haze                                     5
Freezing Rain,Fog                             4
Moderate Snow                                 4
Rain,Snow,Ice Pellets                         4
Freezing Fog                                  4
Snow Showers,Fog                              4
Thunderstorms,Rain Showers,Fog                3
Thunderstorms,Rain                            3
Freezing Drizzle,Haze                         3
Rain,Haze                                     3
Drizzle,Snow                                  2
Moderate Snow,Blowing Snow                    2
Thunderstorms                                 2
Freezing Rain,Haze                            2
Rain Showers,Snow Showers                     2
Freezing Rain,Snow Grains                     1
Thunderstorms,Heavy Rain Showers              1
Thunderstorms,Moderate Rain Showers,Fog       1
Thunderstorms,Rain,Fog                        1
Freezing Rain,Ice Pellets,Fog                 1
Moderate Rain,Fog                             1
Rain,Snow,Fog                                 1
Snow Pellets                                  1
Rain Showers,Fog                              1
Drizzle,Ice Pellets,Fog                       1
Rain,Ice Pellets                              1
Rain,Snow Grains                              1
Name: count, dtype: int64

Powyżej wspomniałem, że `value_counts()` stosuje się do pojedynczej kolumny, ale co, jeśli mamy wymóg uzyskania liczby unikalnych wartości dla podzbioru kolumn lub całej ramki danych.

Istnieją sposoby, dzięki którym możemy rozwiązać ten problem.
* Pętla po każdej kolumnie.
* Użycie `apply` i przekazanie `pd.Series.value_counts`
* itd.

In [32]:
df[['Wind Speed_km/h', 'Weather']].astype(str).apply(pd.Series.value_counts)

Wind Speed_km/h  Weather
0                                                  309.0      NaN
11                                                 791.0      NaN
13                                                 735.0      NaN
15                                                 719.0      NaN
17                                                 666.0      NaN
...                                                  ...      ...
Thunderstorms,Moderate Rain Showers,Fog              NaN      1.0
Thunderstorms,Rain                                   NaN      3.0
Thunderstorms,Rain Showers                           NaN     16.0
Thunderstorms,Rain Showers,Fog                       NaN      3.0
Thunderstorms,Rain,Fog                               NaN      1.0

[84 rows x 2 columns]

# info()
Metoda `info()` w Pandas służy do wyświetlania zwięzłego podsumowania ramki danych (DataFrame). Ta metoda wyświetla podstawowe informacje o ramce danych, w tym:
* zakres indeksu
* typy danych kolumn
* liczbę wartości nie-nullowych
* zużycie pamięci

Wszystkie te szczegóły widzieliśmy powyżej, używając poszczególnych metod lub atrybutów Pandas, jednak za pomocą metody `.info()` możemy je uzyskać w jednej linijce kodu.

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date/Time         8784 non-null   object 
 1   Temp_C            8784 non-null   float64
 2   Dew Point Temp_C  8784 non-null   float64
 3   Rel Hum_%         8784 non-null   int64  
 4   Wind Speed_km/h   8784 non-null   int64  
 5   Visibility_km     8784 non-null   float64
 6   Press_kPa         8784 non-null   float64
 7   Weather           8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


# describe()
Ostatnią, ale nie mniej ważną, jest metoda `describe()` w Pandas, która służy do przeglądania podstawowych szczegółów statystycznych, takich jak percentyle, średnia, odchylenie standardowe itp., dla ramki danych lub serii wartości numerycznych.

Pamiętaj, że metoda `describe()` domyślnie zwraca szczegóły dla wartości numerycznych, jednak możemy również sprawdzić szczegóły dla typu `object`. W takim przypadku metoda zwróci inny wynik, który obejmuje całkowitą liczbę wartości, liczbę unikalnych wartości, najczęściej występującą wartość oraz częstotliwość jej występowania dla każdej kolumny typu `object`.

In [34]:
df.describe()

Temp_C  Dew Point Temp_C    Rel Hum_%  Wind Speed_km/h  \
count  8784.000000       8784.000000  8784.000000      8784.000000   
mean      8.798144          2.555294    67.431694        14.945469   
std      11.687883         10.883072    16.918881         8.688696   
min     -23.300000        -28.500000    18.000000         0.000000   
25%       0.100000         -5.900000    56.000000         9.000000   
50%       9.300000          3.300000    68.000000        13.000000   
75%      18.800000         11.800000    81.000000        20.000000   
max      33.000000         24.400000   100.000000        83.000000   

       Visibility_km    Press_kPa  
count    8784.000000  8784.000000  
mean       27.664447   101.051623  
std        12.622688     0.844005  
min         0.200000    97.520000  
25%        24.100000   100.560000  
50%        25.000000   101.070000  
75%        25.000000   101.590000  
max        48.300000   103.650000

Jeśli przyjrzymy się uważnie wynikowi, łatwo zauważymy, że kolumny `Weather` i `Date/Time` nie są wyświetlane, ponieważ ich typem danych jest `object`.

Wyświetlmy je teraz.

In [35]:
df.describe(include=['object'])

Date/Time       Weather
count              8784          8784
unique             8784            50
top     12/31/2012 7:00  Mainly Clear
freq                  1          2106

Jeśli chcesz zobaczyć statystyki dla wszystkich kolumn, uwzględniając wszystkie typy danych, użyj poniższego kodu.

In [36]:
df.describe(include=['object', 'int', 'float'])

Date/Time       Temp_C  Dew Point Temp_C    Rel Hum_%  \
count              8784  8784.000000       8784.000000  8784.000000   
unique             8784          NaN               NaN          NaN   
top     12/31/2012 7:00          NaN               NaN          NaN   
freq                  1          NaN               NaN          NaN   
mean                NaN     8.798144          2.555294    67.431694   
std                 NaN    11.687883         10.883072    16.918881   
min                 NaN   -23.300000        -28.500000    18.000000   
25%                 NaN     0.100000         -5.900000    56.000000   
50%                 NaN     9.300000          3.300000    68.000000   
75%                 NaN    18.800000         11.800000    81.000000   
max                 NaN    33.000000         24.400000   100.000000   

        Wind Speed_km/h  Visibility_km    Press_kPa       Weather  
count       8784.000000    8784.000000  8784.000000          8784  
unique              NaN            NaN          NaN            50  
top                 NaN            NaN          NaN  Mainly Clear  
freq                NaN            NaN          NaN          2106  
mean          14.945469      27.664447   101.051623           NaN  
std            8.688696      12.622688     0.844005           NaN  
min            0.000000       0.200000    97.520000           NaN  
25%            9.000000      24.100000   100.560000           NaN  
50%           13.000000      25.000000   101.070000           NaN  
75%           20.000000      25.000000   101.590000           NaN  
max           83.000000      48.300000   103.650000           NaN

Metoda `describe()` ma dwa dodatkowe parametry oprócz `include`, a mianowicie `percentile` i `exclude`. Polecam je sprawdzić i zgłębić.

Teraz, gdy znamy już niektóre podstawowe metody i atrybuty, spróbujmy odpowiedzieć na następujące pytania.

Proszę pamiętać, że istnieje wiele sposobów rozwiązania każdego problemu, a tutaj skupiamy się tylko na jednym lub dwóch. Proszę napisz w komentarzach, jeśli użyłeś innej metody, aby inni, w tym ja, mogli z niej skorzystać.

# Pytanie 1: Ustal, ile razy pogoda była "Czysta" (Clear)?

## Podejście 1: value_counts()
Użyj metody `value_counts()`, aby wyświetlić wszystkie zliczenia, a następnie wybierz liczbę dla "Clear".

In [37]:
df['Weather'].value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                              7
Freezing Drizzle,Fog                          6
Snow,Ice Pellets                              6
Snow,Haze                                     5
Freezing Rain,Fog                             4
Moderate Snow                                 4
Rain,Snow,Ice Pellets                         4
Freezing Fog                                  4
Snow Showers,Fog                              4
Thunderstorms,Rain Showers,Fog                3
Thunderstorms,Rain                            3
Freezing Drizzle,Haze                         3
Rain,Haze                                     3
Drizzle,Snow                                  2
Moderate Snow,Blowing Snow                    2
Thunderstorms                                 2
Freezing Rain,Haze                            2
Rain Showers,Snow Showers                     2
Freezing Rain,Snow Grains                     1
Thunderstorms,Heavy Rain Showers              1
Thunderstorms,Moderate Rain Showers,Fog       1
Thunderstorms,Rain,Fog                        1
Freezing Rain,Ice Pellets,Fog                 1
Moderate Rain,Fog                             1
Rain,Snow,Fog                                 1
Snow Pellets                                  1
Rain Showers,Fog                              1
Drizzle,Ice Pellets,Fog                       1
Rain,Ice Pellets                              1
Rain,Snow Grains                              1
Name: count, dtype: int64

Z powyższego wyniku możemy stwierdzić, że pogoda "Clear" wystąpiła 1326 razy.

Jednak nie jest to dobry sposób na pokazanie wyniku.

## Podejście 2: Filtrowanie
Odfiltruj wymagane dane, aby uzyskać liczbę określonej wartości z kolumny.

In [38]:
df['Weather'] == 'Clear'

0       False
1       False
2       False
3       False
4       False
        ...  
8779    False
8780    False
8781    False
8782    False
8783    False
Name: Weather, Length: 8784, dtype: bool

In [39]:
df[df['Weather'] == 'Clear']['Weather'].count()

np.int64(1326)

Idealnie, to jest to, czego szukaliśmy. Zatem 1326 razy pogoda była "Clear".

Jeśli chcesz uzyskać podzbiór danych z pogodą "Clear", po prostu użyj kodu `df[df['Weather'] == 'Clear']`

In [40]:
df[df['Weather'] == 'Clear']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

# Pytanie 2: W które dni prędkość wiatru wynosiła dokładnie 4 km/h i jaka była wtedy pogoda?

In [41]:
df[df['Wind Speed_km/h'] == 4][['Date/Time', 'Weather', 'Wind Speed_km/h']]

Date/Time       Weather  Wind Speed_km/h
0        1/1/2012 0:00           Fog                4
1        1/1/2012 1:00           Fog                4
96       1/5/2012 0:00          Snow                4
101      1/5/2012 5:00          Snow                4
146      1/7/2012 2:00        Cloudy                4
...                ...           ...              ...
8768   12/31/2012 8:00  Snow Showers                4
8769   12/31/2012 9:00          Snow                4
8770  12/31/2012 10:00      Snow,Fog                4
8772  12/31/2012 12:00          Snow                4
8773  12/31/2012 13:00          Snow                4

[474 rows x 3 columns]

Proszę bardzo, jest 474 rekordów, w których prędkość wiatru wynosiła 4 km/h. Mamy datę i godzinę oraz pogodę dla tego okresu.

# Pytanie 3: Podaj listę kolumn, które są typu object.

## Podejście 1: użycie `.dtype` i `.columns`

In [42]:
[c for c in df.columns if df[c].dtype in ['object']]

['Date/Time', 'Weather']

## Podejście 2: użycie alternatywnej metody `.select_dtypes()` i atrybutu `.columns`.

In [43]:
list(df.select_dtypes(['object']).columns)

['Date/Time', 'Weather']